### ISSUE TITLE
addmm for smaller batch slowest than addmm for larger batch with newest cuda libraries

### ISSUE THREAD

**Issue by jsenellart-systran** [torch/torch7#939]
This was observed first in `nn.Linear:updateGradInput` - the issue is probably in CUDA libraries since it shows with 8.0.61 CUDA libraries but not with 8.0.27 - but I am reporting the issue here too for further reference.

when running the script below, we can see that the second operation is ~6 times slower for batch=32 or smaller than for batch=33 or greater. See graph below.

![aa](https://cloud.githubusercontent.com/assets/1520492/23000534/3f578972-f3e1-11e6-8347-c2d8f8b247b8.png)

because of this issue, optimizations based on batch size reduction (for instance adaptive softmax) are totally ruined.

```
require 'nn'
require 'cutorch'

N=nn.Linear(500,50000):cuda()

local T=torch.Timer()

for b=1,64 do
  I=torch.Tensor(b,500):cuda()
  gI=torch.Tensor(b,500):cuda()
  O=torch.Tensor(b,50000):cuda()
  gO=torch.Tensor(b,50000):cuda()

  T:reset()
  for i=1,500 do
    -- from nn.Linear:updateOutput
    O:addmm(0, O, 1, I, N.weight:t())
  end
  cutorch.synchronize()
  local uo=T:time().real

  T:reset()
  for i=1,500 do
    -- from nn.Linear:updateGradInput
    gI:addmm(0,gI,1,gO,N.weight)
  end
  cutorch.synchronize()
  print(b,uo,T:time().real)
end
```

**Comment by vince62s** [comment]
Trying to revive this issue ...
I am not 100% sure it's only a cuda issue.
When running the code on CPU only it is also 2X slower with batch size <=31 versus batsh size >=32

```
require 'nn'
--require 'cutorch'

N=nn.Linear(500,50000)
--:cuda()

local T=torch.Timer()

for b=30,33 do
  I=torch.Tensor(b,500)
--:cuda()
  gI=torch.Tensor(b,500)
--:cuda()
  O=torch.Tensor(b,50000)
--:cuda()
  gO=torch.Tensor(b,50000)
--:cuda()

  T:reset()
  for i=1,500 do
    -- from nn.Linear:updateOutput
    O:addmm(0, O, 1, I, N.weight:t())
  end
--  cutorch.synchronize()
  local uo=T:time().real

  T:reset()
  for i=1,500 do
    -- from nn.Linear:updateGradInput
    gI:addmm(0,gI,1,gO,N.weight)
  end
--  cutorch.synchronize()
  print(b,uo,T:time().real)
end
```
```
th test.lua
30	23.32201218605	24.21106505394	
31	24.783236026764	25.714804887772	
32	19.756242990494	11.576460123062	
33	20.319017887115	12.197923898697	
```

**Comment by scnakandala** [comment]
Hi,
Were you able to find a fix for this problem? I am facing similar slowdowns the fully-connected layers of AlexNet from addmm operations compared to conv layers. Fully-connected layers should be around 5-10% of the total runtime. But right now they are consuming ~50% with small batch sizes.
